File: fix_imports.py
import os
import re

def process_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    # Find package and imports
    package_match = re.search(r'^package\s+[\w\.]+;', content, re.MULTILINE)
    
    # We want to find fully qualified names like com.example.zhanfinancebackend.modules.auth.entity.User
    # but NOT in package or import statements.
    
    # regex to find com.example.zhanfinancebackend...ClassName
    # It must end with a Capitalized word (the class name)
    # We must exclude lines starting with "import " or "package "
    
    lines = content.split('\n')
    new_lines = []
    imports_to_add = set()
    
    fqn_pattern = re.compile(r'\b(com\.example\.zhanfinancebackend\.[\w\.]+\.([A-Z]\w*))\b')
    
    modified = False
    
    for line in lines:
        if line.strip().startswith('import ') or line.strip().startswith('package '):
            new_lines.append(line)
            continue
            
        # find all FQNs in the line
        matches = fqn_pattern.findall(line)
        if matches:
            for fqn, class_name in matches:
                # exclude if it's somehow just a package or if the FQN doesn't actually have a class
                imports_to_add.add(fqn)
                line = line.replace(fqn, class_name)
                modified = True
                
        new_lines.append(line)
        
    if not modified:
        return False
        
    # Now we need to insert the new imports right after the package statement or existing imports
    # Let's collect existing imports to avoid duplicates
    existing_imports = set(re.findall(r'^import\s+([\w\.]+);', content, re.MULTILINE))
    
    final_imports = imports_to_add - existing_imports
    
    if not final_imports:
        # We replaced FQNs but imports were already there!
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write('\n'.join(new_lines))
        return True
        
    # Find where to insert
    insert_idx = 0
    last_import_idx = -1
    package_idx = -1
    
    for i, line in enumerate(new_lines):
        if line.startswith('package '):
            package_idx = i
        elif line.startswith('import '):
            last_import_idx = i
            
    if last_import_idx != -1:
        insert_idx = last_import_idx + 1
    elif package_idx != -1:
        insert_idx = package_idx + 1
    else:
        insert_idx = 0
        
    # Insert imports
    imports_str = [f"import {imp};" for imp in sorted(final_imports)]
    
    # add an empty line before and after if needed
    if insert_idx > 0 and new_lines[insert_idx-1].strip() != '':
        imports_str.insert(0, "")
        
    # Check if we have multiple classes with the same name!
    # If so, we CANNOT import both. We must keep one fully qualified or it breaks.
    # Let's check for clashes.
    class_names = [imp.split('.')[-1] for imp in (final_imports | existing_imports)]
    if len(class_names) != len(set(class_names)):
        print(f"Skipping {filepath} due to class name clash in imports!")
        return False
        
    for i, imp_str in enumerate(imports_str):
        new_lines.insert(insert_idx + i, imp_str)
        
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write('\n'.join(new_lines))
        
    return True

def main():
    root_dir = r"c:\Users\murat\IdeaProjects\JF-1C\zhan-finance-backend\src\main\java\com\example\zhanfinancebackend"
    modified_count = 0
    for subdir, _, files in os.walk(root_dir):
        for file in files:
            if file.endswith('.java'):
                filepath = os.path.join(subdir, file)
                if process_file(filepath):
                    modified_count += 1
                    print(f"Updated {filepath}")
                    
    print(f"Total files updated: {modified_count}")

if __name__ == "__main__":
    main()

File: fix_imports.js
const fs = require('fs');
const path = require('path');

function processFile(filepath) {
    let content = fs.readFileSync(filepath, 'utf8');

    // Find any occurrence of com.example.zhanfinancebackend.something
    const prefixRegex = /\b(com\.example\.zhanfinancebackend\.[\w\.]+)\b/g;

    const lines = content.split('\n');
    let modified = false;
    let newLines = [];
    let importsToAdd = new Set();
    
    for (let line of lines) {
        if (line.trim().startsWith('import ') || line.trim().startsWith('package ')) {
            newLines.push(line);
            continue;
        }

        let lineCopy = line;
        let match;
        let matches = [];
        while ((match = prefixRegex.exec(lineCopy)) !== null) {
            matches.push(match[1]);
        }

        // Sort by length descending to replace longer FQNs first if any overlap
        matches.sort((a, b) => b.length - a.length);

        for (let fqn of matches) {
            // Split FQN by dot
            let parts = fqn.split('.');
            let packageParts = [];
            let classParts = [];
            let foundClass = false;

            for (let part of parts) {
                if (!foundClass && /^[A-Z]/.test(part)) {
                    foundClass = true;
                }
                if (foundClass) {
                    classParts.push(part);
                } else {
                    packageParts.push(part);
                }
            }

            if (classParts.length > 0) {
                let classFqn = packageParts.join('.') + '.' + classParts[0];
                let replacement = classParts.join('.');
                
                importsToAdd.add(classFqn);
                
                // Replace this specific fqn occurrence
                lineCopy = lineCopy.split(fqn).join(replacement);
                modified = true;
            }
        }
        
        newLines.push(lineCopy);
    }

    if (!modified) {
        return false;
    }

    // Extract existing imports
    const existingImports = new Set();
    const existingImportRegex = /^import\s+([\w\.]+);/gm;
    let importMatch;
    while ((importMatch = existingImportRegex.exec(content)) !== null) {
        existingImports.add(importMatch[1]);
    }

    const finalImports = [...importsToAdd].filter(imp => !existingImports.has(imp));

    if (finalImports.length === 0) {
        // Just replaced inline usage, no new imports
        if (content !== newLines.join('\n')) {
             fs.writeFileSync(filepath, newLines.join('\n'), 'utf8');
             return true;
        }
        return false;
    }

    // Check for class name clashes
    const allClassNames = [...existingImports, ...finalImports].map(imp => imp.split('.').pop());
    const uniqueClassNames = new Set(allClassNames);
    if (allClassNames.length !== uniqueClassNames.size) {
        console.log(`Skipping ${filepath} due to class name clash in imports!`);
        return false;
    }

    // Find insertion index
    let insertIdx = 0;
    let lastImportIdx = -1;
    let packageIdx = -1;

    for (let i = 0; i < newLines.length; i++) {
        if (newLines[i].startsWith('package ')) packageIdx = i;
        else if (newLines[i].startsWith('import ')) lastImportIdx = i;
    }

    if (lastImportIdx !== -1) insertIdx = lastImportIdx + 1;
    else if (packageIdx !== -1) insertIdx = packageIdx + 1;
    else insertIdx = 0;

    finalImports.sort();
    const importStrings = finalImports.map(imp => `import ${imp};`);
    
    if (insertIdx > 0 && newLines[insertIdx - 1].trim() !== '') {
        importStrings.unshift('');
    }

    newLines.splice(insertIdx, 0, ...importStrings);

    fs.writeFileSync(filepath, newLines.join('\n'), 'utf8');
    return true;
}

function walkDir(dir, callback) {
    fs.readdirSync(dir).forEach(f => {
        let dirPath = path.join(dir, f);
        let isDirectory = fs.statSync(dirPath).isDirectory();
        isDirectory ? walkDir(dirPath, callback) : callback(path.join(dir, f));
    });
}

function main() {
    const rootDir = path.join(__dirname, 'zhan-finance-backend', 'src', 'main', 'java', 'com', 'example', 'zhanfinancebackend');
    let modifiedCount = 0;
    
    walkDir(rootDir, function(filePath) {
        if (filePath.endsWith('.java')) {
            if (processFile(filePath)) {
                console.log(`Updated ${filePath}`);
                modifiedCount++;
            }
        }
    });

    console.log(`Total files updated: ${modifiedCount}`);
}

main();
